# 4. Project in Applied Computational Multi-Omics
##### _MCMSc in Computational Biology and Bioinformatics, June 2024_
###### _Authors: Beatriz Infante, Inês Pacheco and Maria Inês Gomes_

## Cross results from RNA-Seq and ChiP-Seq Analysis

1. **Annotate ChIPseq peaks**

Map the ChIPseq peaks to closest genes using R packages (ChIPpeakAnno e GenomicsRanges). You can follow the example here: https://bioinformaticscore-shared-training.github.io/cruk-summer-school2019/ChIPSeq/Materials/Practicals/Day4/Practical4_Downstream_Analysis_of_ ChIP-seq_SS.html#chip-peak-annotation


```{r}
peakfile <-read.table("SOX15_peaks_filtered.bed")
head(peakfile)
```

Get a 400bp window around summit (-200 and +200)
```{r}
Motif_Peaks <- data.frame((peakfile$V1), (peakfile$V2-200), (peakfile$V2+200), stringsAsFactors = FALSE)
head(Motif_Peaks)
```

Create a bed file and list files in directory and check if the file you created is there.
```{r}
options(scipen=999)

write.table(Motif_Peaks, file= "Peaks_for_motif_detection.bed", 
            row.names=FALSE,col.names = FALSE, sep="\t", quote =FALSE)

dir()
```

Read the bed file and convert the peaks to a GRanges object
```{r}
df1<-read.table("Peaks_for_motif_detection.bed", header=FALSE)

gr1 <- GRanges(seqnames=df1$V1, ranges=IRanges(start=df1$V2, end=df1$V3))
```
Create annotation file from EnsDb (Ensembl) or TxDb (transcript annotation) packages
```{r}
annoData <- toGRanges(EnsDb.Hsapiens.v86, feature="gene")
annoData[1:2]
```
Annotate the peak GRanges object with peaks mapped to gene with a -2000 and 500 bp window around the TSS
```{r}
anno.gr1 <- annotatePeakInBatch(gr1, AnnotationData=annoData, output="nearestBiDirectionalPromoters", bindingRegion=c(-4000, 500))
```
Trim out of bound ranges, annotate with Entrez IDs and list annotated peaks
```{r}
anno.gr1 <- trim(anno.gr1)
anno.gr1 <- addGeneIDs(anno.gr1,"org.Hs.eg.db",IDs2Add = "entrez_id")
head(anno.gr1)
length(unique(anno.gr1$gene_name))
```
#### Identify direct targets of your TF.
Cross results from RNAseq and ChIPseq analyses in R/Python to identify the genes directly regulated by your Transcription Factor. DEGs.xlsx
```{r}
library(readxl)
DEG <- read_excel("path/to/your/folder")
```
```{r}
DEG <- as.data.frame(DEG)
rownames(DEG) <- DEG[,1]
DEG <- DEG[,-1]
head(DEG)
```

Cross results from RNA-Seq and ChiP-Seq Analysis
```{r}
intersect(x = row.names(DEG), y = (anno.gr1$gene_name))
```

